### PR TITLE
HTTP read optimization

### PR DESCRIPTION
**Pull request by samansmink**
This PR adds some optimizations to both parquet reading and CSV reading. 

## Before this PR
Before this PR the CSV and Parquet reads over HTTP go through the HTTPFilesystem. To ensure the amount of requests stays sane when small reads are made, HTTP requests currently always fetch 1 MB of data, which is buffered in the HTTPFilesystem basically functioning as a Read Ahead buffer for small reads.

The CSV reader itself also contains a buffer where between 16KB and 2MB bytes are buffered, depending on the CSV line size. The CSV reader is single threaded and reads synchronously.

The parquet reader reads through the HTTPFS abstraction as well, but reads multithreaded where each thread gets assigned a number of rowgroups to scan. Because of the way the HTTPFilesystem fetches data in 1MB chunks, a certain amount of "overscanning" occurs where data is read which is not used.

## Optimizations
The optimizations to both the CSV and parquet reader aim to make the respective reader aware of the fact that its writing to a high latency filesystem and adjust their prefetching/buffering strategy accordingly.

### CSV Reader
- Circumvent HTTPFS buffering, directly read into buffer from CSV reader
- increase buffer size to 32MB (in parsing mode only) to reduce request count
- do async read aheads (in parsing mode only)

### Parquet Reader
- Add prefetch mechanism to prefetch columnchunks to scan efficiently and reduce overscanning
	- For full scan: Parquet reader will determine size of row group and prefetch the whole row group in a single request
	- For partial scan: The parquet reader will register the bytes ranges it will need for scanning a row group. Adjacent ranges are merged if they lie within 16kb of eachother. These ranges are then read directly, bypassing the HTTPFileystem buffering.

- New skipping optimization:
	- Parquet skips are done lazily: only on a read call to a column reader, the skips are performed
	- The prefetches on these columns are also done lazily: only when the columnchunk is touched, the prefetch for the whole columnchunk will happen. This can prevent columnchunks from being scanned at all when filters on other columns have not filtered out all tuples in the columnchunk.

### CSV
There is a ~16x reduction in requests for large files.
Speedup will vary wildly depending on the type of network and system.

note however that parsing CSV is actually pretty slow still, we should probably consider multithreaded CSV parsing.

### Parquet
**Full scans**
For most files there will not be a big difference, for example `CREATE TABLE test as SELECT * FROM 's3://test-bucket/tpch_lineitem_sf10.csv';` has a 2% reduction in data transferred. However there is ~5x reduction in request count (from 2991 to 600).

When row groups are very small ( smaller than the 1MB HTTPFS read size), the full scan performance gain from this optimization will be a bit higher, but this is probably not very common.

**Partial scans**
For partial scans and complex queries we do get a decent reduction in transfered bytes. To demonstrate this I've run all queries from TPCH TPCDS and counted the requests and bytes transferred over the network.

TPCH SF1
branch | bytes downloaded | request count 
--- | --- | ---
master | 3.87GB | 4081
optimized | 1.29GB | 4368

TPC-DS SF1
branch | bytes downloaded | request count 
-- | -- | --
master | 12.40GB | 15787
optimized | 2.39GB | 26182

### A note on row group size
The row group size is very important in these experiments. The current row group size in the duckdb parquet reader is 100k. The TPC-H and TPC-DS results in the benchmark are somewhat inflated by DuckDB's current row group size. To demonstrate this, here are the results for the TPC-DS on row group size of 1M:

TPC-DS SF1 1M Row groups
branch | bytes downloaded | request count 
-- | -- | --
master | 4.51GB | 7746
optimized | 2.29GB | 7006

Note the significantly reduced request size and also the better performance for the current master branch. Also note that while in some cases, the optimizations in this PR will actually increase the request count, this can easily be resolved by increasing the row group size in duckdb, which can be done by the user already.

### Skipping
Query from test/sql/copy/parquet/parquet_http_prefetch.test
```
CREATE TABLE test_fetch_delay (a INT, b INT);
INSERT INTO test_fetch_delay (SELECT (i%2) * 2, (i%2) * 2 from range(0,2500000) as tbl(i));
COPY test_fetch_delay to 's3://test-bucket/skip_delay.parquet';
CREATE TABLE test as SELECT * from 's3://test-bucket/skip_delay.parquet' where a = 1;
```
Reduced bytes read : 945KB to 474KB (2x)

**TPCH and TPCDS**
The Skipping does not really make a difference on average, but does increase the total request count by about 30%

## Next steps
- Analyze this new optimization in DuckDB WASM: we probably want to circumvent the existing RA buffer there with the DIRECT_IO flag.
- improve CSV reader to support multithreaded reads
- improve parquet skips even further, if we find out use cases where the skipping optimization makes a big difference, for example in queries with several low selectivity filters and many columns scanned, there are ideas to use the adaptive filters from @taniabogatsch

### CODE REVIEW

**Comment by samansmink** [comment]
note: 2 failing tests are unrelated to this PR

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

File: extension/parquet/parquet_reader.cpp [hunk 1]
```diff
@@ -563,12 +624,20 @@ void ParquetReader::InitializeScan(ParquetReaderScanState &state, vector<column_
 	state.sel.Initialize(STANDARD_VECTOR_SIZE);
 
 	if (!state.file_handle || state.file_handle->path != file_handle->path) {
-		state.file_handle =
-		    file_handle->file_system.OpenFile(file_handle->path, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK,
-		                                      FileSystem::DEFAULT_COMPRESSION, file_opener);
+		auto flags = FileFlags::FILE_FLAGS_READ;
+
+		if (file_handle->file_system.GetName() == "HTTPFileSystem") {
```
[the remark this answers is not in the document]
Comment (samansmink) [reply extension/parquet/parquet_reader.cpp]: Good point! Changed it to a combination of the existing functions: !file_handle->OnDiskFile() && file_handle->CanSeek()

**Comment by samansmink** [comment]
@Mytherin as we discussed i've reverted the "fake-async io" code in the csv reader to a super simple approach that just uses a bigger buffer when the filesystem indicates that its not an on disk file.

**Comment by zenazn** [comment]
Really excited for this one! :)

**Comment by Mytherin** [comment]
Thanks for the updates! Looks good to me.

**Pull request merged**